text/0000-uninstall-node.md

```diff
@@ -0,0 +1,64 @@
+- Feature Name: uninstall_node
+- Start Date: 2024-08-21
+- RFC PR:[https://github.com/volta-cli/volta/pull/1882]
+- Volta Issue:
+
+# Summary
+
+[summary]: #summary
+
+Add support for uninstalling node to `volta uninstall`
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, Volta doesn't support uninstalling node due to this main reasons(as discussed in [#327](!https://github.com/volta-cli/volta/issues/327)): Its main purpose is to save disk space, nowadays disk size is not a big problem in many cases. So this task is in a low priority.But community has this demand since 2019, they need to manually remove some files to achieve that.Add support for uninstalling node is more convinent in this case, btw since we have `install` here we should also have `uninstall` here.
+
+# Pedagogy
+
+[pedagogy]: #pedagogy
+
+The user's intuition and way of treating it are opposite to `install`. Since `install` fetch node image, unpack it and set it as default, `unintall` should remove those things.
+
+# Details
+
+[details]: #details
+
+The new command will have the structure:
+
+```
+volta uninstall node@version
+```
+
+totally the same with `install`.
+
+Some examples:
+
+```
+volta uninstall node@20.16.0
+```
+
+```
+volta uninstall node@lts
+```
+
+By add a `uninstall` to `Tool` trait and move uninstall logic from `Spec` to `Node, Package, Npm, Yarn, Pnpm`, it would resolve version with `node::resolve` first, and then remove tar.gz file and npm file in `home.node_inventory_dir()` and node in `home.node_image_root_dir()`
+
+# Critique
+
+[critique]: #critique
+
+In fact we want `uninstall` to save disk space or just to remove those things we don't need , it seems like a command like `prune` or `clean` is more suitable, which tracks every tool's status, and when `prune` is runned, we can clean all those tools that is no longer used.
+
+In this case, maybe we should use a file to track every tool. When `volta pin` or `volta install`
+is runned, the toolchain must be added to that file (means this toolchain is used in a project or is set as default), and removed outdated toolchain from that file. And we can delete all that we don't use anymore.
+
+But this way will make some big change in volta, and how to track a toolchain's status is complex.
+
+# Unresolved questions
+
+[unresolved]: #unresolved-questions
+
+1. What should `uninstall` do when uninstall a default node?
+   Since `install` actually fetch (in background),unpack(in background), and set as default(in user's view), `uninstall` should at least remove tar.gz file, npm file and node dir. but what `uninstall` should do when uninstall a node version that is set as default?
```

Comment:
I think this is an important question to resolve, though it's tricky with the overloaded behavior of `install` that you noted. I wonder if it would be clearer as a user with the proposed change to use `pin` for everything. If we did that, then we could have a command `unpin` that explicitly removes the pinned version, while `clean` or `prune` would remove the cached versions. That would clearly separate the two cases while making it more clear what each does.

Reply:
You mean ```volta remove``` just focus on disk cleaning and ```volta unpin ``` focus on deleting default version message(in package.json)? It looks nice.
So how about this: When user is running ```volta remove``` to remove default node version(no matter pinned in a project or set as global default one( These two cases all mean that user is trying to remove a in-using node)), we info user that they are removing a default one from disk and  breake the removing process.

Reply:
@charlespierce @karthest Yeah, I think `unpin` is *distinctly* the wrong verb for *removing it from the file system*. It should exactly and only, well, un-pin it, i.e. make it not-used in the current project or as a user default, at least in my view. Then, as you say, it should be a *separate* verb to actually clear the cached download. (I at least would find it *astonishing* if it deleted it.) But at that point should it even… be a thing? I cannot see a case where I would actually want to `volta unpin` something, only cases where I would want to `volta prune` or `volta uninstall` where `uninstall` means “get it off my system” rather than “remove it from the config”. 🤔 I think, anyway. 🤔